### PR TITLE
de.ient.RDPlot

### PR DESCRIPTION
**Pull request by nolyn**
I think this should include everything. I tried for days to get PyQt5 to work, by building it from source, but pip never accepted the built module. I tried various configs  and qt versions from other peoples manifests. Finally I ended up using the wheel, which works. Hope that's ok.

### CODE REVIEW

**Comment by TingPing** [comment]
> Finally I ended up using the wheel, which works. Hope that's ok.

Its not, it bundles massive libs and links against who knows what ABI. Others have built it I think @bochecha recently built it if he has any advice.

File: de.ient.RDPlot.appdata.xml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  <name>RDPlot</name>
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license> GPL-3.0-or-later</project_license>
+  <developer_name>Lehrstuhl und Institut für Nachrichtentechnik, RWTH Aachen</developer_name>
+  <summary>RDPlot is a tool for plotting rate distortion curves</summary>
+  <description>
+    <p>RD Plot is a plotting tool for visualizing rate-distortion data parsed from
+encoder log files of common video coding reference software packages.
+It can:
+    Display rate-distortion plots
```
Comment (TingPing) [inline de.ient.RDPlot.appdata.xml]: These should probably be in an `<ul>` rather than just shoved in a paragraph.

File: de.ient.RDPlot.yaml [hunk 2]
```diff
@@ -0,0 +1,162 @@
+app-id: de.ient.RDPlot
+runtime: org.kde.Platform
+runtime-version: '5.10'
```
Comment (TingPing) [inline de.ient.RDPlot.yaml]: Choose 5.9 for lts or 5.11 for latest version ideally.

File: de.ient.RDPlot.yaml [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+build-options:
+  append-path: "/usr/lib/sdk/gfortran-62/bin"
+modules:
+  - name: gfortran
+    buildsystem: simple
+    build-commands:
+    - "/usr/lib/sdk/gfortran-62/install.sh"
+
+  - name: lapack
+    buildsystem: cmake
```
Comment (TingPing) [inline de.ient.RDPlot.yaml]: `cmake-ninja`

File: de.ient.RDPlot.yaml [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    sources:
+    - type: archive
+      url: http://www.netlib.org/lapack/lapack-3.8.0.tar.gz
+      sha512: 17786cb7306fccdc9b4a242de7f64fc261ebe6a10b6ec55f519deb4cb673cb137e8742aa5698fd2dc52f1cd56d3bd116af3f593a01dcf6770c4dcc86c50b2a7f
+    cleanup:
+    - "/lib/cmake"
+  - name: python3-numpy
+    buildsystem: simple
+    build-commands:
+    - python3 setup.py build -j 0
```
Comment (TingPing) [inline de.ient.RDPlot.yaml]: `-j 0` should probably be `-j $FLATPAK_BUILDER_N_JOBS`

File: de.ient.RDPlot.yaml [hunk 5]
```diff
@@ -0,0 +1,162 @@
+app-id: de.ient.RDPlot
```
Comment (TingPing) [inline de.ient.RDPlot.yaml]: Please follow this style guide: https://github.com/flathub/flathub/wiki/YAML-Style-Guide

File: de.ient.RDPlot.yaml (hunk 3 above)
Comment (nolyn) [reply de.ient.RDPlot.yaml]: this yields
```
CMake Error at CMakeLists.txt:3 (project):
  The Ninja generator does not support Fortran using Ninja version
```

Comment (TingPing) [reply de.ient.RDPlot.yaml]: Ok, fine to keep as is then.

File: de.ient.RDPlot.yaml (hunk 4 above)
Comment (nolyn) [reply de.ient.RDPlot.yaml]: thx, changed it. can't do the same for scipy, apperently it cannot be build multi-core

File: de.ient.RDPlot.appdata.xml [hunk 6]
```diff
@@ -11,12 +11,12 @@
     <p>RD Plot is a plotting tool for visualizing rate-distortion data parsed from
 encoder log files of common video coding reference software packages.
 It can:
-    Display rate-distortion plots
-    Select component to be plotted (YUV, Y, U, V and even more)
-    Plot rd plot over time (rate over time, psnr over time)
-    Exporting plots (through matplotlib interface)
-    Calculate and display table of bjontegaard delta measurement
-    Export tables as latex style table code
+    <ul>Display rate-distortion plots</ul>
```
Comment (TingPing) [inline de.ient.RDPlot.appdata.xml]: This isn't valid. For one use `appstream-util validate`.

Anyway its html style syntax:
```html
<p>It supports these features:</p>
<ul>
  <li>Does cool thing #1</li>
  <li>Does cool thing #2</li>
</ul>
```

File: de.ient.RDPlot.yaml [hunk 7]
```diff
@@ -15,55 +15,57 @@ modules:
...
+      - type: archive
+        url: https://pypi.python.org/packages/0b/66/86185402ee2d55865c675c06a5cfef742e39f4635a4ce1b1aefd20711c13/numpy-1.14.2.zip
+        sha256: facc6f925c3099ac01a1f03758100772560a0b020fb9d70f210404be08006bcb
+
   - name: python3-scipy
     buildsystem: simple
     build-commands:
-    - python3 setup.py build -j 0
-    - python3 setup.py install --prefix=/app --root=/ --optimize=1
+      - python3 setup.py build -j 0
```
Comment (TingPing) [inline de.ient.RDPlot.yaml]: Same `-j` comment.

File: de.ient.RDPlot.yaml (hunk 4 above)
Comment (TingPing) [reply de.ient.RDPlot.yaml]: Alright, was worth a shot anyway.

**Comment by bochecha** [comment]
PyQt5 is a bit hard indeed…

See here: https://github.com/flathub/org.musicbrainz.Picard/blob/master/org.musicbrainz.Picard.json#L48-L66

You'll probably also want sip: https://github.com/flathub/org.musicbrainz.Picard/blob/master/org.musicbrainz.Picard.json#L29-L47

**Comment by nolyn** [comment]
I tried that, except the patch, but that seems not to be used in the manuscript. The build runs through, but when I try to run it afterwards it fails. Any idea what the problem might be?:
```
Traceback (most recent call last):
  File "/app/bin/rdplot", line 5, in <module>
    from pkg_resources import load_entry_point
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 634, in _load_backward_compatible
  File "/usr/lib/python3.5/site-packages/setuptools-22.0.5-py3.5.egg/pkg_resources/__init__.py", line 2927, in <module>
  File "/usr/lib/python3.5/site-packages/setuptools-22.0.5-py3.5.egg/pkg_resources/__init__.py", line 2913, in _call_aside
  File "/usr/lib/python3.5/site-packages/setuptools-22.0.5-py3.5.egg/pkg_resources/__init__.py", line 2940, in _initialize_master_working_set
  File "/usr/lib/python3.5/site-packages/setuptools-22.0.5-py3.5.egg/pkg_resources/__init__.py", line 635, in _build_master
  File "/usr/lib/python3.5/site-packages/setuptools-22.0.5-py3.5.egg/pkg_resources/__init__.py", line 943, in require
  File "/usr/lib/python3.5/site-packages/setuptools-22.0.5-py3.5.egg/pkg_resources/__init__.py", line 829, in resolve
pkg_resources.DistributionNotFound: The 'pyqt5<5.11' distribution was not found and is required by rdplot
```

**Comment by eszlari** [comment]
I think you need to set `PYTHONPATH`, see:

https://github.com/flathub/org.kde.krita/blob/master/org.kde.krita.yaml

**Comment by nolyn** [comment]
I tried that, but I still get the same error. I added 
```
finish-args:
  - --env=PYTHONPATH=/app/lib/python3.5/site-packages
build-options:
  env:
    PYTHON: "/usr/bin/python3"
    PIP: "/usr/bin/pip3"
    PYTHONPATH: /app/lib/python3.5/site-packages	
```
Also I can do
```
cd /var/lib/flatpak/app/de.ient.RDPlot/current/active/files/lib/python3.5/site-packages/
ipython
Python 3.7.0 (default, Jul 15 2018, 10:44:58) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.5.0 -- An enhanced Interactive Python. Type '?' for help.
In [1]: import PyQt5
In [2]: PyQt5.__file__
Out[2]: '/var/lib/flatpak/app/de.ient.RDPlot/x86_64/master/e60735ece44adaa7f830924e368c72d480fc4786de3d55d8b5de9d3673f78ef9/files/lib/python3.5/site-packages/PyQt5/__init__.py'
```
using my system intepreter. I can also do it from the buildDir:
```
flatpak-builder --run buildDir de.ient.RDPlot.yaml bash                                                                         :(
$ cd /app/lib/python3.5/site-packages/
$ python3.5
Python 3.5.2 (default, Jan 10 2018, 06:17:28) 
[GCC 6.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import PyQt5
>>> PyQt5.__file__
'/app/lib/python3.5/site-packages/PyQt5/__init__.py'
```

**Comment by nolyn** [comment]
some more digging shows that pip does not know the package is there, after all it was not installed by pip
```
flatpak-builder --run buildDir de.ient.RDPlot.yaml bash                                                                           :(
$ pip3 list
atomicwrites (1.1.5)
attrs (18.1.0)
cycler (0.10.0)
gitdb2 (2.0.4)
GitPython (2.1.11)
jsonpickle (0.9.6)
kiwisolver (1.0.1)
matplotlib (2.2.3)
matplotlib2tikz (0.6.17)
meson (0.46.0)
more-itertools (4.3.0)
mpldatacursor (0.6.2)
numpy (1.14.2)
pathlib2 (2.3.2)
Pillow (5.2.0)
pip (8.1.2)
pluggy (0.7.1)
py (1.5.4)
pyparsing (2.2.0)
pytest (3.7.1)
python-dateutil (2.7.3)
pytz (2018.5)
rdplot (1.2.6)
scandir (1.9.0)
scipy (1.0.1)
setuptools (22.0.5)
setuptools-scm (3.1.0)
sip (4.19.12)
six (1.11.0)
smmap2 (2.0.4)
tabulate (0.8.2)
xmltodict (0.11.0)
```
wonder whether that can be related, though I don't see how yet. since the module is there and can be imported

**Comment by nolyn** [comment]
part of the problem is that I had pyqt and sip in install_requires:
```
install_requires=['cycler', 'matplotlib', 'numpy', 'py', 'pyparsing', 'pyqt5<5.11', 'pytest', 'python-dateutil', 'pytz',
                 'sip', 'six', 'scipy', 'tabulate', 'mpldatacursor',
```
things in install_requires should usually be source distributions, but those are wheels. if I remove them I can launch the app.
However, now I have new problem:
```
sudo flatpak install --reinstall RDPlot-repo de.ient.RDPlot.Debug
flatpak run --devel --command=sh --system-talk-name=org.freedesktop.login1 de.ient.RDPlot                                       :(
sh-4.3$ rdplot
Gtk-Message: Failed to load module "canberra-gtk-module"
Gtk-Message: Failed to load module "canberra-gtk-module"
Qt: Session management error: Could not open network socket
Added sub class '<class 'rdplot.SimulationDataItemClasses.DatLogs.DatLogJEM70_360'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.DatLogs.DatLogJEM501_360'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.DatLogs.DatLogConversionPSNRLoss360'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.DatLogs.DatLogJEM70'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.DatLogs.DatLogVTM1_0'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItem.AbstractSimulationDataItem'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.DatLogs.DatLogJEM70Geo'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.DatLogs.DatLogBasedOnClassName'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.DatLogs.DatLogHEVC'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.DatLogs.DatLogJEM70GeoHOMC'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.DatLogs.AbstractDatLog'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.DecoderAnalyserLogs.AbstractDecAnalyserLog'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.DecoderAnalyserLogs.DecAnalyserLogHM'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItem.AbstractSimulationDataItem'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.EncoderLogs.EncLogHM'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.EncoderLogs.EncLogSHM'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.EncoderLogs.EncLogHM360Lib'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItemClasses.EncoderLogs.AbstractEncLog'>' to simulation data item  factory
Added sub class '<class 'rdplot.SimulationDataItem.AbstractSimulationDataItem'>' to simulation data item  factory
libGL error: MESA-LOADER: failed to retrieve device information
libGL error: Version 4 or later of flush extension not found
libGL error: failed to load driver: i915
libGL error: failed to open drm device: No such file or directory
libGL error: failed to load driver: i965
```
Opening a file, which works in non flatpak version:
```
Parsed '/home/sauer/Software/rd-plot-gui/examplEncLogs/HM-16.6-JEM-2.0-GeoV7_PPP_PMVP-BasketballDrill_832x480_50_encoder+lowdelay+P+jvet10_FTBE100_QP27_enc.log' 
ASSERT: "last >= first" in file itemmodels/qabstractitemmodel.cpp, line 2743
Aborted (core dumped)
```

how can I debug it? Ideally with Pycharm, but other debuggers are fine too. There is no pdb available and gdb doesn't work either:
```
flatpak run --devel --command=sh --system-talk-name=org.freedesktop.login1 de.ient.RDPlot                                       :(
sh-4.3$ gdb python
GNU gdb (GDB) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-unknown-linux".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Traceback (most recent call last):
  File "/home/sauer/Software/gdb-imagewatch/install/bin/gdb-imagewatch/gdb-imagewatch.py", line 66, in <module>
    main()
  File "/home/sauer/Software/gdb-imagewatch/install/bin/gdb-imagewatch/gdb-imagewatch.py", line 59, in main
    window = giwwindow.GdbImageWatchWindow(script_path, debugger)
  File "/home/sauer/Software/gdb-imagewatch/install/bin/gdb-imagewatch/giwscripts/giwwindow.py", line 35, in __init__
    script_path + '/libgiwwindow.so')
  File "/usr/lib/python3.5/ctypes/__init__.py", line 425, in LoadLibrary
    return self._dlltype(name)
  File "/usr/lib/python3.5/ctypes/__init__.py", line 347, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libpython3.6m.so.1.0: cannot open shared object file: No such file or directory
Reading symbols from python...(no debugging symbols found)...done.
```
why does it try to use python 3.6 here?

**Comment by nolyn** [comment]
bot, build

**Comment by nolyn** [comment]
After a long while I gave this another try. I was able to build a working version locally. The problems disappeared after using different pyqt sip and platform versions.

**Comment by nedrichards** [comment]
bot, build de.ient.RDPlot
testbot, build de.ient.RDPlot

**Comment by nolyn** [comment]
hope i can do this too ....
bot, build de.ient.RDPlot
testbot, build de.ient.RDPlot

**Comment by nolyn** [comment]
i don't think i can ... please trigger the build again?

**Comment by nolyn** [comment]
ping @nedrichards

**Comment by nolyn** [comment]
why is no one looking at this? because of the failing label or this pull request is so old? should i close this PR and make a new one?

its running on my laptop already using flatpak-builder, the only remaining thing I have to fix is a issue that the version can not be determined correctly when building on flathub. and I think it should work now, but needs to be tested/built of course ...

**Comment by aviwad** [comment]
> i don't think i can ... please trigger the build again?

they had turned the infrastructure off temporarily to upgrade it to the newer one. test bot works again now. also they haven't merged my pull request now either, I guess it's because they (the maintainers) do this in their free time :D

**Comment by nolyn** [comment]
alright. thanks for telling me this. i hope i can trigger a build then

**Comment by nolyn** [comment]
testbot, build de.ient.RDPlot

**Comment by aviwad** [comment]
odd, test bot worked for me a day ago. ask on the #flatpak irc chat on freenode?

**Comment by nedrichards** [comment]
testbot, build de.ient.RDPlot
bot, build de.ient.RDPlot

**Comment by nolyn** [comment]
downloading some sources failed. worked before, I didn't change that. it also doesn't work when I do it manually
`
wget https://downloads.sourceforge.net/projects/pyqt/sip/sip-4.19.13/sip-4.19.13.tar.gz
--2019-02-16 19:04:22--  https://downloads.sourceforge.net/projects/pyqt/sip/sip-4.19.13/sip-4.19.13.tar.gz
Loaded CA certificate '/etc/ssl/certs/ca-certificates.crt'
Resolving downloads.sourceforge.net (downloads.sourceforge.net)... 216.105.38.13
Connecting to downloads.sourceforge.net (downloads.sourceforge.net)|216.105.38.13|:443... connected.
HTTP request sent, awaiting response... 307 Temporary Redirect
Location: https://sourceforge.net/downloads [following]
--2019-02-16 19:04:28--  https://sourceforge.net/downloads
Resolving sourceforge.net (sourceforge.net)... 216.105.38.13
Connecting to sourceforge.net (sourceforge.net)|216.105.38.13|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://sourceforge.net/publish [following]
--2019-02-16 19:04:34--  https://sourceforge.net/publish
Reusing existing connection to sourceforge.net:443.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://sourceforge.net/publish/ [following]
--2019-02-16 19:04:35--  https://sourceforge.net/publish/
Reusing existing connection to sourceforge.net:443.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://sourceforge.net/create [following]
--2019-02-16 19:04:35--  https://sourceforge.net/create
Reusing existing connection to sourceforge.net:443.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://sourceforge.net/create/ [following]
--2019-02-16 19:04:36--  https://sourceforge.net/create/
Reusing existing connection to sourceforge.net:443.
HTTP request sent, awaiting response... 200 OK
Length: 139060 (136K) [text/html]
Saving to: ‘sip-4.19.13.tar.gz’

sip-4.19.13.tar.gz                                   100%[=====================================================================================================================>] 135.80K   157KB/s    in 0.9s    

2019-02-16 19:04:38 (157 KB/s) - ‘sip-4.19.13.tar.gz’ saved [139060/139060]
`
the file I get is not a tar.gz. and the checksum is wrong. 
do i have to use a different sip version? i used this combination of sip and pyqt, since I saw it to be working for others. I think I took the snipped from https://github.com/flathub/org.kde.krita

when I download it manually from the sourceforge website, the link i get is https://sourceforge.net/projects/pyqt/files/sip/sip-4.19.13/sip-4.19.13.tar.gz/download and the checksum is correct.

**Comment by nedrichards** [comment]
bot, build de.ient.RDPlot

**Comment by nolyn** [comment]
it worked :-). I'll do some more cleanup, restrict the architecture to x86_64 and then it should be ready ...

**Comment by nedrichards** [comment]
Great! fwiw there are new features available for developers at flathub today, so you should be able to run the test build yourself when that's done https://blogs.gnome.org/alexl/2019/02/19/changes-in-flathub-land/

**Comment by nolyn** [comment]
testbot, build de.ient.RDPlot
bot, build de.ient.RDPlot

**Comment by nolyn** [comment]
it doesnt build for i386, but for x86_64, arm, aarch64. arm and aarch64 will be mostly smartphones? i don't think it would make sense there. so i'll just keep x86_64

**Comment by nolyn** [comment]
bot, build de.ient.RDPlot

**Comment by nolyn** [comment]
bot, build de.ient.RDPlot

**Comment by nedrichards** [comment]
arm certainly is laptops from my former employers, endless, raspberry pi etc. generally we encourage having all architectures on unless there's a specific reason.

**Comment by nolyn** [comment]
I want to rename the app to match our other one: https://github.com/flathub/de.rwth_aachen.ient.YUView.
So from de.ient.RDPlot to de.rwth_aachen.ient.RDPlot

didn't work though. should I just create a new branch? wasn't necessary the other time though I think.

**Comment by nolyn** [comment]
i'll try doing a different branch with the correct name and making a new PR

**Comment by nolyn** [comment]
continued in PR #873

**Pull request closed without merging**